warehouse/forklift/legacy.py

```diff
@@ -572,6 +572,14 @@ def _is_valid_dist_file(filename, filetype):
     return True
 
 
+def _is_duplicate_file(request, filename, sha256_hash):
+    """ Check to see if file already exists, and if it's content matches
+    """
+    requested_file = request.db.query(File).filter(File.filename == filename)
+    return (request.db.query(requested_file.exists()).scalar() and
```

Comment:
We can probably avoid issuing two queries by just loading the existing file from the database using something like:

```python
file_ = (
    request.db.query(File)
              .filter(File.filename == filename)
              .first()
)

return file_ is None or file_.sha256_digest == sha256_hash
```

If we want to minimize what data we pull from the DB, we could change the query to be:

```python
file_ = (
    request.db.query(File)
              .options(load_only("sha256_digest"))
              .filter(File.filename == filename)
              .first()
)
```

Reply:
Ooo, love removing redundant queries. Also, I was wondering if we'd want to check all the hashes like we do later in this code path., that makes lots of sense. Thanks for the feedback! Give me a bit and I'll fix this up.

Reply:
By the way, if you're not aware you can click the little red box on the right hand side of the screen to get taken to the debug toolbar, and you can see all the queries that got executed as part of a HTTP request. It records the last N HTTP requests so you can still view the ones for file upload, you just need to load the page in the browser and then navigate to the upload request in the debug toolbar.

I normally like to do that to try and look for queries that seem close enough to duplicates I might be able to share a single query for, and also when using strategies like ``load_only()`` to make sure I'm not accidently triggering extra queries because I forgot something in my ``load_only()`` (since SQLAlchemy will implicitly execute extra queries if you say, did the ``load_only("sha256_digest")`` from above, and then accessed ``file_.size``).

Reply:
That's an A+ tip, I completely forgot about the toolbar.  This will help tremendously.